Allow DDP to handle mixed CPU-GPU models

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32779 Allow DDP to handle mixed CPU-GPU models**

Differential Revision: [D19625701](https://our.internmc.facebook.com/intern/diff/D19625701)

mrshenli committed:
Allow DDP to handle mixed CPU-GPU models

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit fc09ee84:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=fc09ee8404d2a4dbe939aff4703c115f6b874b3e).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4371311) pytorch_linux_xenial_cuda10_1_cudnn7_py3_multigpu_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92728817) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92728817)) :snowflake:
<details>
<summary>
<code>Jan 29 20:14:40 AssertionError: 10 not less than or equal to 1e-05 :</code>
</summary>

```
Jan 29 20:14:40 ---------------------------------------------------------------------- 
Jan 29 20:14:40 Traceback (most recent call last): 
Jan 29 20:14:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 174, in wrapper 
Jan 29 20:14:40     self._join_processes(fn) 
Jan 29 20:14:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 255, in _join_processes 
Jan 29 20:14:40     self._check_return_codes(elapsed_time) 
Jan 29 20:14:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 279, in _check_return_codes 
Jan 29 20:14:40     self.assertEqual(first_process.exitcode, 0) 
Jan 29 20:14:40   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 892, in assertEqual 
Jan 29 20:14:40     super(TestCase, self).assertLessEqual(abs(x - y), prec, message) 
Jan 29 20:14:40 AssertionError: 10 not less than or equal to 1e-05 :  
Jan 29 20:14:40  
Jan 29 20:14:40 ---------------------------------------------------------------------- 
Jan 29 20:14:40 Ran 118 tests in 216.446s 
Jan 29 20:14:40  
Jan 29 20:14:40 FAILED (failures=1, skipped=4) 
Jan 29 20:14:40  
Jan 29 20:14:40 Generating XML reports... 
Jan 29 20:14:41 Traceback (most recent call last): 
Jan 29 20:14:41   File "test/run_test.py", line 456, in <module> 
Jan 29 20:14:41     main() 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

mrshenli committed:
Update on "Allow DDP to handle mixed CPU-GPU models"


Differential Revision: [D19625701](https://our.internmc.facebook.com/intern/diff/D19625701)

[ghstack-poisoned]
